contrib/envoyproxy/go-control-plane/cmd/serviceextensions/main.go

```diff
@@ -0,0 +1,162 @@
+	extproc.ExternalProcessorServer
+}
+
+type serviceExtensionConfig struct {
+	extensionPort   string
+	extensionHost   string
+	healthcheckPort string
+}
+
+func loadConfig() serviceExtensionConfig {
+	extensionPortInt := internal.IntEnv("DD_SERVICE_EXTENSION_PORT", 443)
+	if extensionPortInt < 1 || extensionPortInt > 65535 {
+		log.Error("service_extension: invalid port number: %d\n", extensionPortInt)
+		os.Exit(1)
+	}
+
+	healthcheckPortInt := internal.IntEnv("DD_SERVICE_EXTENSION_HEALTHCHECK_PORT", 80)
+	if healthcheckPortInt < 1 || healthcheckPortInt > 65535 {
+		log.Error("service_extension: invalid port number: %d\n", healthcheckPortInt)
+		os.Exit(1)
+	}
+
+	extensionHost := internal.IpEnv("DD_SERVICE_EXTENSION_HOST", net.IP{0, 0, 0, 0})
+	extensionPortStr := strconv.FormatInt(int64(extensionPortInt), 10)
+	healthcheckPortStr := strconv.FormatInt(int64(healthcheckPortInt), 10)
+
+	return serviceExtensionConfig{
+		extensionPort:   extensionPortStr,
+		extensionHost:   extensionHost.String(),
+		healthcheckPort: healthcheckPortStr,
+	}
+}
+
+func main() {
+	var extensionService AppsecCalloutExtensionService
+
+	// Set the DD_VERSION to the current tracer version if not set
+	if os.Getenv("DD_VERSION") == "" {
+		if err := os.Setenv("DD_VERSION", version.Tag); err != nil {
+			log.Error("service_extension: failed to set DD_VERSION environment variable: %v\n", err)
+		}
+	}
+
+	config := loadConfig()
+
+	tracer.Start(tracer.WithAppSecEnabled(true))
+	// TODO: Enable ASM standalone mode when it is developed (should be done for Q4 2024)
+
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	g, ctx := errgroup.WithContext(ctx)
+
+	g.Go(func() error {
+		return startGPRCSsl(ctx, &extensionService, config)
+	})
+
+	g.Go(func() error {
+		return startHealthCheck(ctx, config)
+	})
+
+	if err := g.Wait(); err != nil {
+		log.Error("service_extension: %v\n", err)
+	}
+
+	log.Info("service_extension: shutting down\n")
+	cancel()
+	tracer.Stop()
+}
+
+func startHealthCheck(ctx context.Context, config serviceExtensionConfig) error {
+	muxServer := mux.NewRouter()
+	muxServer.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte(`{"status": "ok", "library": {"language": "golang", "version": "` + version.Tag + `"}}`))
+	})
+
+	server := &http.Server{
+		Addr:    config.extensionHost + ":" + config.healthcheckPort,
+		Handler: muxServer,
+	}
+
+	go func() {
+		<-ctx.Done()
+		shutdownCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		if err := server.Shutdown(shutdownCtx); err != nil {
+			log.Error("service_extension: health check server shutdown: %v\n", err)
+		}
+	}()
+
+	log.Info("service_extension: health check server started on %s:%s\n", config.extensionHost, config.healthcheckPort)
+	if err := server.ListenAndServe(); err != nil {
```

Comment:
you might want to check here if `errors.Is(err, http. ErrServerClosed)`, which is what is returned when you call `server.Shutdown` and it should be expected so no need to return an error.

Reply:
Applied in e3cc085 😄